### ISSUE TITLE
Classifier LibSVM_SVC classifier never yields results

### ISSUE THREAD

**Issue by engelen** [automl/auto-sklearn#425]
Branch: develop.

It appears that, when using the `LibSVM_SVC`, no valid results are ever obtained. The classifier always crashed when being `fit`. Consider this minimal test case:
```python
import numpy as np
import sklearn.model_selection
import sklearn.datasets
import sklearn.metrics
from sklearn.datasets import fetch_mldata

iris = fetch_mldata('iris')

# Split labeled data into training and test set
X_train, X_test, y_train, y_test = \
    sklearn.model_selection.train_test_split(iris.data, iris.target, random_state=1)

# Run auto-sklearn
automl = autosklearn.classification.AutoSklearnClassifier(
    initial_configurations_via_metalearning=0,
    time_left_for_this_task=30, per_run_time_limit=15,
    tmp_folder='./autoslearn_cv_example_tmp',
    output_folder='./autosklearn_cv_example_out',
    delete_tmp_folder_after_terminate=False,
    include_estimators=['libsvm_svc'])

automl.fit(X_train.copy(), y_train.copy(), dataset_name='iris')

# Refit to full data
automl.refit(X_train.copy(), y_train.copy())

# Predict on test data and show performance
predictions = automl.predict(X_test)
print("Accuracy score", sklearn.metrics.accuracy_score(y_test, predictions))

# Output models
print(automl.show_models())
```

this will output the following (where the accuracy, of course, has mean 1/3):
```
Accuracy score 0.342105263158
[(1.000000, MyDummyClassifier(configuration=1, init_params=None, random_state=None)),
]
```

The full error trace is as follows:
```
[DEBUG] [2018-02-13 09:25:52,384:AutoMLSMBO(1)::iris] Return: Status: <StatusType.CRASHED: 3>, cost: 1.000000, time: 2.087767, additional: {'configuration_origin': 'Random Search (sorted)', 'traceback': 'Traceback (most recent call last):
  File "{local-path}/auto-sklearn/autosklearn/evaluation/__init__.py", line 25, in fit_predict_try_except_decorator
    return ta(queue=queue, **kwargs)
  File "{local-path}/auto-sklearn/autosklearn/evaluation/train_evaluator.py", line 437, in eval_holdout
    evaluator.fit_predict_and_loss(iterative=iterative)
  File "{local-path}/auto-sklearn/autosklearn/evaluation/train_evaluator.py", line 105, in fit_predict_and_loss
    i, train_indices=train_split, test_indices=test_split
  File "{local-path}/auto-sklearn/autosklearn/evaluation/train_evaluator.py", line 288, in _partial_fit_and_predict
    self.Y_train[train_indices])
  File "{local-path}/auto-sklearn/autosklearn/evaluation/abstract_evaluator.py", line 372, in _fit_and_suppress_warnings
    model.fit(X, y)
  File "{local-path}/auto-sklearn/autosklearn/pipeline/base.py", line 93, in fit
    self.fit_estimator(X, y, **fit_params)
  File "{local-path}/auto-sklearn/autosklearn/pipeline/base.py", line 110, in fit_estimator
    self._final_estimator.fit(X, y, **fit_params)
  File "{local-path}/auto-sklearn/autosklearn/pipeline/components/base.py", line 380, in fit
    return self.choice.fit(X, y, **kwargs)
  File "{local-path}/auto-sklearn/autosklearn/pipeline/components/classification/libsvm_svc.py", line 75, in fit
    self.estimator.fit(X, Y)
  File "/anaconda3/envs/askl-dev/lib/python3.5/site-packages/sklearn/svm/base.py", line 187, in fit
    fit(X, y, sample_weight, solver_type, kernel, random_seed=seed)
  File "/anaconda3/envs/askl-dev/lib/python3.5/site-packages/sklearn/svm/base.py", line 254, in _dense_fit
    max_iter=self.max_iter, random_seed=random_seed)
  File "sklearn/svm/libsvm.pyx", line 187, in sklearn.svm.libsvm.fit
ValueError: cache_size <= 0
', 'error': "ValueError('cache_size <= 0',)"}
```

**Comment by mfeurer** [comment]
I cannot reproduce this. Are you on a MAC?

**Comment by engelen** [comment]
I am on a Mac indeed. Apparently, somebody else encountered this issue as well some time ago (on a Mac): https://github.com/automl/auto-sklearn/issues/253#issuecomment-287679886.

I've found the cause of the problem: https://github.com/automl/auto-sklearn/blob/master/autosklearn/pipeline/components/classification/libsvm_svc.py#L37. In my case, `maxrss` is greater than `soft`, and `cache_size` becomes negative:
```
Soft: 3072
MaxRSS: 54468
cache_size: -34264
```

**Comment by mfeurer** [comment]
Hm, this should not happen. Do you maybe want to provide a fix for this?

**Comment by engelen** [comment]
(sorry for the unintentional closing)

Sure. Could you provide some inline commenting for the current code? Specifically, what's being calculated and how.

**Comment by mfeurer** [comment]
Sure. The SVC in scikit-learn has a kernel cache to fit large kernels into RAM without eating all RAM. As Auto-sklearn uses a memory limit, we try to dynamically allocate a certain percentage (66%) of that limit to the kernel cache. Therefore, we first figure out how much memory we use (maxrss) and the limit (soft) to then calculate the size (line 37). However, if maxrss is larger than soft, this will fail. So maybe there should be a single if statement after the if/else before the except to catch this issue?

**Comment by engelen** [comment]
Thanks. An alternative to the current approach would be to use [`psutil`](https://github.com/giampaolo/psutil), which works on 
![image](https://user-images.githubusercontent.com/2271433/36259301-17143ba8-125e-11e8-9c8e-167ffd72cbf6.png)
Do you see a reason not to do this, @mfeurer? It would also [remove the dependency on the `resource` module](https://github.com/automl/auto-sklearn/issues/427).

**Comment by mfeurer** [comment]
`resource` is not only used in the SVM part of Auto-sklearn, but also for limiting the memory and runtime of the machine learning algorithm. Can psutil limit the memory and runtime of a subprocess? Only then it would be a full alternative.

**Comment by engelen** [comment]
Hmm, is `resource` used in an `auto-sklearn` dependency, then? In the `auto-sklearn` package itself, it appears to be only imported in
- `pipeline/components/classification/libsvm_svc.py`
- `pipeline/components/regression/libsvm_svr.py`

**Comment by mfeurer** [comment]
`resource` is used by the [pynisher](https://pypi.python.org/pypi/pynisher) which limits the memory and runtime usage of the individual models. Note to myself: we need to document this!

**Comment by engelen** [comment]
The root cause appears to be the fact that `resource.getrusage` returns the RSS in bytes on MacOS, and in kilobytes on other platforms. Interesting. See http://fa.bianp.net/blog/2013/different-ways-to-get-memory-consumption-or-lessons-learned-from-memory_profiler/.

I opted for staying with the `resource` module in the pull request (https://github.com/automl/auto-sklearn/pull/430) to change as little as possible. I also added a check for `cache_size < 0`.

Note: should `resource` ever be removed as a dependency from `auto-sklearn`'s dependencies, 
 `process.memory_info().rss` can be considered as an alternative.
